sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,432 @@
+    if (columnarBatch != null) {
+      columnarBatch.close()
+      columnarBatch = null
+    }
+    if (rows != null) {
+      rows.close()
+      rows = null
+    }
+  }
+
+  /**
+   * Initialize ORC file reader and batch record reader.
+   * Please note that `setRequiredSchema` is needed to be called after this.
+   */
+  override def initialize(inputSplit: InputSplit, taskAttemptContext: TaskAttemptContext): Unit = {
+    val fileSplit = inputSplit.asInstanceOf[FileSplit]
+    val conf = taskAttemptContext.getConfiguration
+    reader = OrcFile.createReader(
+      fileSplit.getPath,
+      OrcFile.readerOptions(conf)
+        .maxLength(OrcConf.MAX_FILE_LENGTH.getLong(conf))
+        .filesystem(fileSplit.getPath.getFileSystem(conf)))
+
+    val options = OrcInputFormat.buildOptions(conf, reader, fileSplit.getStart, fileSplit.getLength)
+    rows = reader.rows(options)
+  }
+
+  /**
+   * Set required schema and partition information.
+   * With this information, this creates ColumnarBatch with the full schema.
+   */
+  def setRequiredSchema(
+      orcSchema: TypeDescription,
+      requestedColIds: Array[Int],
+      resultSchema: StructType,
+      requiredSchema: StructType,
+      partitionValues: InternalRow): Unit = {
+    batch = orcSchema.createRowBatch(DEFAULT_SIZE)
+    assert(!batch.selectedInUse)
+    totalRowCount = reader.getNumberOfRows
+    logDebug(s"totalRowCount = $totalRowCount")
+
+    this.requiredSchema = requiredSchema
+    this.requestedColIds = requestedColIds
+
+    val memMode = DEFAULT_MEMORY_MODE
+    val capacity = ColumnarBatch.DEFAULT_BATCH_SIZE
+    if (memMode == MemoryMode.OFF_HEAP) {
+      columnVectors = OffHeapColumnVector.allocateColumns(capacity, resultSchema)
+    } else {
+      columnVectors = OnHeapColumnVector.allocateColumns(capacity, resultSchema)
+    }
+    columnarBatch = new ColumnarBatch(resultSchema, columnVectors.toArray, capacity)
+
+    if (partitionValues.numFields > 0) {
+      val partitionIdx = requiredSchema.fields.length
+      for (i <- 0 until partitionValues.numFields) {
+        ColumnVectorUtils.populate(columnVectors(i + partitionIdx), partitionValues, i)
+        columnVectors(i + partitionIdx).setIsConstant()
+      }
+    }
+  }
+
+  /**
+   * Return true if there exists more data in the next batch. If exists, prepare the next batch
+   * by copying from ORC VectorizedRowBatch columns to Spark ColumnarBatch columns.
+   */
+  private def nextBatch(): Boolean = {
+    if (rowsReturned >= totalRowCount) {
+      return false
+    }
+
+    rows.nextBatch(batch)
+    val batchSize = batch.size
+    if (batchSize == 0) {
+      return false
+    }
+    rowsReturned += batchSize
+    columnarBatch.reset()
+    columnarBatch.setNumRows(batchSize)
+
+    var i = 0
+    while (i < requiredSchema.length) {
+      val field = requiredSchema(i)
+      val toColumn = columnVectors(i)
+
+      if (requestedColIds(i) < 0) {
+        toColumn.appendNulls(batchSize)
```

Comment:
The column ids not requested are fixed. We don't need to reset them and re-append nulls into them again for each batch. We only need to append nulls into them once.

Reply:
`The column ids not requested are fixed` is done by OrcDeserializer.

Reply:
I meant, the columns not requested are not changed for each batch. For now every time you reset the batch and then re-append nulls into such columns. But those columns can be reused.

Once #20116 is merged, we reset by column vectors, not the columnar batch. So we can just reset the column vectors of the required columns.

Reply:
I see, @viirya .